File: 374-guess-number-higher-or-lower/BinarySearch.java
/**
 * 二分查找（bisectLeft）：
 * 如果目标值更小(或相等)，左边；否则右边
 */
public class Solution extends GuessGame {
  public int guessNumber(int n) {
      return bisectLeft(n);
  }
  
  private int bisectLeft (int n) {
      int left = 1;
      int right = n;
      
      while (left < right) {
          int middle = left + (right - left) / 2;
          if (guess(middle) == 1) { // 目标数字比较大
              left = middle + 1;
          } else {
              right = middle;
          }
      }
      
      return left;
  }
}

File: problems/278-first-bad-version/binary-search.js
/**
 * 二分查找：
 * 如果middle是bad的，那么源头肯定在左边（包括当前位置）；否则在右边。
 * 使用bisectLeft
 */
var solution = function (isBadVersion) {

  function bisectLeft (n) {
    let [left, right] = [1, n]

    while (left < right) {
      let middle = left + Math.floor((right - left) / 2)
      if (isBadVersion(middle)) {
        right = middle
      } else {
        left = middle + 1
      }
    }

    return left
  }

  return function (n) {
    return bisectLeft(n)
  };
};
